docs/src/userguide/source_files_and_compilation.rst

```diff
@@ -854,6 +854,14 @@ Cython code.  Here is the list of currently supported directives:
     signature, which cannot otherwise be retrieved after
     compilation.  Default is False.
 
+``embedsignature.pure`` (True / False)
+    If set to True, Cython will do a its best attempt to use
+    pure-Python type annotations in embedded signatures. If arguments
+    have explicit type annotations, these are preferred over C
+    types. Otherwise, C types are mapped to the closest Python type
+    equivalent (e.g., C ``short`` is mapped to Python ``int`` type and
+    C ``double`` is mapped to Python ``float`` type).  Default is False.
+
```

Comment:
I never considered having both a C type declaration and a Python type declaration for the same argument. Not sure if it's a good idea to allow (and even encourage) that. I can see in your test examples that it works (apparently), and I can see it being useful in order to disambiguate C type declarations. However, we're currently trying to discourage C type declarations in favour of type annotations in Python syntax, and this counters that.

```suggestion
``embedsignature.pure`` (True / False)
    If set to True, Cython will do its best attempt to use
    pure-Python type annotations in embedded signatures.  If arguments
    have explicit Python type annotations, these are preferred over C type
    declarations.  Otherwise, declared C types are mapped to the closest
    Python type equivalent (e.g., C ``short`` is mapped to Python ``int`` type
    and C ``double`` is mapped to Python ``float`` type). 
    Implies ``embedsignature=True``.  Default is False.

```

Reply:
Remember that "Python type declarations" are actually just annotations. 

If you want to reject arguments with both C types and annotations, you will make my life miserable, I'll have a lot of work to do in mpi4py to accomodate for that change, to the point that you would force me to remove all C types form arguments of `def` functions. 

Moreover, your call will prevent users for doing some very reasonable stuff, For example

```python
def foo(int arg: Annotated[int, MaxValueAllowed(42)]) -> Any:
    ...
```